Wrappers/Python/ccpi/optimisation/functions/Function.py

```diff
@@ -23,55 +23,493 @@
+        
+        Due to Moreau’s identity, we have an analytic formula to compute the proximal operator of the convex conjugate :math:`F^{*}`
+        
+        .. math:: \mathrm{prox}_{\tau F^{*}}(x) = x - \tau\mathrm{prox}_{\tau^{-1} F}(\tau^{-1}x)
+                
+        """
+        if out is None:
+            return x - tau * self.proximal(x/tau, 1/tau)
+        else:            
+            self.proximal(x/tau, 1/tau, out = out)
+            out*=-tau
+            out.add(x, out = out)                       
 
     def grad(self, x):
-        '''Alias of gradient(x,None)'''
+        """Alias of gradient(x,None)"""
         warnings.warn('''This method will disappear in following 
         versions of the CIL. Use gradient instead''', DeprecationWarning)
         return self.gradient(x, out=None)
 
     def prox(self, x, tau):
-        '''Alias of proximal(x, tau, None)'''
+        """ Alias of proximal(x, tau, None)"""
         warnings.warn('''This method will disappear in following 
         versions of the CIL. Use proximal instead''', DeprecationWarning)
         return self.proximal(x, tau, out=None)
+    
+    def domain(self):
+        """  Returns the domain of the function :math:`F`. 
+        Although, it is mathematically correct to setup a function based on the domain, 
+        we do not use this information at the moment. """
+        return self.domain
+    
+    # Algebra for Function Class
+    
+        # Add functions
+        # Subtract functions
+        # Add/Substract with Scalar
+        # Multiply with Scalar
+    
+    def __add__(self, other):
+        
+        """ Returns the sum of the functions.
+        
+            Cases: a) the sum of two functions :math:`(F_{1}+F_{2})(x) = F_{1}(x) + F_{2}(x)`
+                   b) the sum of a function with a scalar :math:`(F_{1}+scalar)(x) = F_{1}(x) + scalar`
 
-    def __rmul__(self, scalar):
-        '''Defines the multiplication by a scalar on the left
+        """
+        
+        if isinstance(other, Function):
+            return SumFunction(self, other)
+        elif isinstance(other, (SumFunctionScalar, ConstantFunction, Number)):
+            return SumFunctionScalar(self, other)
+        else:
+            raise ValueError('Not implemented')   
+            
+    def __radd__(self, other):
+        
+        """ Making addition commutative. """
+        return self + other 
+                          
+    def __sub__(self, other):
+        """ Returns the subtraction of the functions."""
+        return self + (-1) * other    
 
-        returns a ScaledFunction'''
+    def __rmul__(self, scalar):
+        """Returns a function multiplied by a scalar."""               
         return ScaledFunction(self, scalar)
+    
+    def __mul__(self, scalar):
+        """ Returns a function multiplied by a scalar from the left."""                    
+        return scalar * ScaledFunction(self, 1)   
+    
+    def centered_at(self, center):
```

Comment:
Have we agreed on this name `centered_at`?

Reply:
Not sure, it's not a bad name and as you pointed out translate maybe confusing...At the end it will depend on the documentation